### ISSUE TITLE
What does Grad mean when training Demucs

### ISSUE THREAD

**Issue by lyndonlauder** [facebookresearch/demucs#315]
Hi

I did a practice training session with a small part of my dataset. Everything is going well but at Epoch 37 the train loss went up and the Grad number went up a lot. Please can you explain to me what Grad means and why this jump happened? The train loss and grad went back down afterwards.

Thanks!

[04-04 22:52:33][demucs.solver][INFO] - Training...
[04-04 22:54:07][demucs.solver][INFO] - Train | Epoch 35 | 83/332 | 0.89 it/sec | Loss 0.1098 | Reco 0.1098 | Grad 0.0700
[04-04 22:55:40][demucs.solver][INFO] - Train | Epoch 35 | 166/332 | 0.89 it/sec | Loss 0.1094 | Reco 0.1094 | Grad 0.0725
[04-04 22:57:13][demucs.solver][INFO] - Train | Epoch 35 | 249/332 | 0.89 it/sec | Loss 0.1121 | Reco 0.1121 | Grad 0.0744
[04-04 22:58:45][demucs.solver][INFO] - Train Summary | Epoch 35 | Loss=0.1127 | Reco=0.1127 | Grad=0.0732
[04-04 22:58:45][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 22:58:45][demucs.solver][INFO] - Cross validation...
[04-04 22:58:46][demucs.solver][INFO] - Valid Summary | Epoch 35 | Loss=0.1552 | Reco=0.1552 | Nsdr=6.484 | Best=0.1501 | Bname=main
[04-04 22:58:54][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 22:58:54][demucs.solver][INFO] - Training...
[04-04 23:00:28][demucs.solver][INFO] - Train | Epoch 36 | 83/332 | 0.89 it/sec | Loss 0.1114 | Reco 0.1114 | Grad 0.0690
[04-04 23:02:00][demucs.solver][INFO] - Train | Epoch 36 | 166/332 | 0.89 it/sec | Loss 0.1128 | Reco 0.1128 | Grad 0.0700
[04-04 23:03:33][demucs.solver][INFO] - Train | Epoch 36 | 249/332 | 0.89 it/sec | Loss 0.1123 | Reco 0.1123 | Grad 0.0705
[04-04 23:05:05][demucs.solver][INFO] - Train Summary | Epoch 36 | Loss=0.1117 | Reco=0.1117 | Grad=0.0706
[04-04 23:05:05][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 23:05:05][demucs.solver][INFO] - Cross validation...
[04-04 23:05:06][demucs.solver][INFO] - Valid Summary | Epoch 36 | Loss=0.1546 | Reco=0.1546 | Nsdr=6.550 | Best=0.1501 | Bname=main
[04-04 23:05:10][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 23:05:10][demucs.solver][INFO] - Training...
[04-04 23:06:44][demucs.solver][INFO] - Train | Epoch 37 | 83/332 | 0.89 it/sec | Loss 0.1138 | Reco 0.1138 | Grad 0.0662
[04-04 23:08:17][demucs.solver][INFO] - Train | Epoch 37 | 166/332 | 0.89 it/sec | Loss 0.1321 | Reco 0.1321 | Grad 1.9693
[04-04 23:09:50][demucs.solver][INFO] - Train | Epoch 37 | 249/332 | 0.89 it/sec | Loss 0.1367 | Reco 0.1367 | Grad 1.3576
[04-04 23:11:22][demucs.solver][INFO] - Train Summary | Epoch 37 | Loss=0.1369 | Reco=0.1369 | Grad=1.0501
[04-04 23:11:22][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 23:11:22][demucs.solver][INFO] - Cross validation...
[04-04 23:11:24][demucs.solver][INFO] - Valid Summary | Epoch 37 | Loss=0.1539 | Reco=0.1539 | Nsdr=6.560 | Best=0.1501 | Bname=main
[04-04 23:11:27][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 23:11:27][demucs.solver][INFO] - Training...
[04-04 23:13:01][demucs.solver][INFO] - Train | Epoch 38 | 83/332 | 0.89 it/sec | Loss 0.1288 | Reco 0.1288 | Grad 0.0917
[04-04 23:14:34][demucs.solver][INFO] - Train | Epoch 38 | 166/332 | 0.89 it/sec | Loss 0.1279 | Reco 0.1279 | Grad 0.0871
[04-04 23:16:07][demucs.solver][INFO] - Train | Epoch 38 | 249/332 | 0.89 it/sec | Loss 0.1282 | Reco 0.1282 | Grad 0.0834
[04-04 23:17:39][demucs.solver][INFO] - Train Summary | Epoch 38 | Loss=0.1267 | Reco=0.1267 | Grad=0.0815
[04-04 23:17:39][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 23:17:39][demucs.solver][INFO] - Cross validation...
[04-04 23:17:40][demucs.solver][INFO] - Valid Summary | Epoch 38 | Loss=0.1549 | Reco=0.1549 | Nsdr=6.527 | Best=0.1501 | Bname=main
[04-04 23:17:44][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 23:17:44][demucs.solver][INFO] - Training...
[04-04 23:19:18][demucs.solver][INFO] - Train | Epoch 39 | 83/332 | 0.89 it/sec | Loss 0.1230 | Reco 0.1230 | Grad 0.0750
[04-04 23:20:51][demucs.solver][INFO] - Train | Epoch 39 | 166/332 | 0.89 it/sec | Loss 0.1222 | Reco 0.1222 | Grad 0.0707
[04-04 23:22:24][demucs.solver][INFO] - Train | Epoch 39 | 249/332 | 0.89 it/sec | Loss 0.1204 | Reco 0.1204 | Grad 0.0726
[04-04 23:23:56][demucs.solver][INFO] - Train Summary | Epoch 39 | Loss=0.1184 | Reco=0.1184 | Grad=0.0716
[04-04 23:23:56][demucs.solver][INFO] - ----------------------------------------------------------------------
[04-04 23:23:56][demucs.solver][INFO] - Cross validation...
[04-04 23:23:58][demucs.solver][INFO] - Valid Summary | Epoch 39 | Loss=0.1546 | Reco=0.1546 | Nsdr=6.531 | Best=0.1501 | Bname=main

**Comment by adefossez** [comment]
grad is the total norm of the gradient over all layers. this can be useful to spot instabilities or divergence issues.
In your case it does seem a bit unstable, looking at epoch 37 for instance, with the training loss going up and gradient loss becoming quite large. You can set gradient clipping to limit the issue (`optim.clip_grad`). in your case I would set it to 0.2.
Alternatively, or complementarily you could decrease the value of `optim.beta2` from 0.999 to 0.99 or 0.9.
